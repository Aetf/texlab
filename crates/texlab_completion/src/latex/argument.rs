use super::combinators::{self, Parameter};
use crate::factory;
use async_trait::async_trait;
use std::iter;
use texlab_feature::{FeatureProvider, FeatureRequest};
use texlab_protocol::{CompletionItem, CompletionParams, TextEdit};

#[derive(Debug, PartialEq, Eq, Clone, Copy, Default)]
pub struct LatexArgumentCompletionProvider;

#[async_trait]
impl FeatureProvider for LatexArgumentCompletionProvider {
    type Params = CompletionParams;
    type Output = Vec<CompletionItem>;

    async fn execute<'a>(&'a self, req: &'a FeatureRequest<Self::Params>) -> Self::Output {
        let mut all_items = Vec::new();
        for comp in req.view.components() {
            for cmd in &comp.commands {
                let name = format!("\\{}", cmd.name);
                for (i, param) in cmd.parameters.iter().enumerate() {
                    let mut items = combinators::argument(
                        req,
                        iter::once(Parameter {
                            name: &name,
                            index: i,
                        }),
                        |ctx| async move {
                            let mut items = Vec::new();
                            for arg in &param.0 {
                                let text_edit = TextEdit::new(ctx.range, (&arg.name).into());
                                let item = factory::argument(
                                    req,
                                    &arg.name,
                                    text_edit,
                                    arg.image.as_ref().map(AsRef::as_ref),
                                );
                                items.push(item);
                            }
                            items
                        },
                    )
                    .await;
                    all_items.append(&mut items);
                }
            }
        }
        all_items
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;
    use texlab_feature::FeatureTester;
    use texlab_protocol::{CompletionTextEditExt, Range, RangeExt};

    #[tokio::test]
    async fn empty_latex_document() {
        let actual_items = FeatureTester::new()
            .file("main.tex", "")
            .main("main.tex")
            .position(0, 0)
            .test_completion(LatexArgumentCompletionProvider)
            .await;

        assert!(actual_items.is_empty());
    }

    #[tokio::test]
    async fn empty_bibtex_document() {
        let actual_items = FeatureTester::new()
            .file("main.bib", "")
            .main("main.bib")
            .position(0, 0)
            .test_completion(LatexArgumentCompletionProvider)
            .await;

        assert!(actual_items.is_empty());
    }

    #[tokio::test]
    async fn inside_mathbb_empty() {
        let actual_items = FeatureTester::new()
            .file(
                "main.tex",
                indoc!(
                    r#"
                        \usepackage{amsfonts}
                        \mathbb{}
                    "#
                ),
            )
            .main("main.tex")
            .position(1, 8)
            .test_completion(LatexArgumentCompletionProvider)
            .await;

        assert!(!actual_items.is_empty());
        assert_eq!(
            actual_items[0]
                .text_edit
                .as_ref()
                .and_then(|edit| edit.text_edit())
                .map(|edit| edit.range)
                .unwrap(),
            Range::new_simple(1, 8, 1, 8)
        );
    }

    #[tokio::test]
    async fn inside_mathbb_non_empty() {
        let actual_items = FeatureTester::new()
            .file(
                "main.tex",
                indoc!(
                    r#"
                        \usepackage{amsfonts}
                        \mathbb{foo}
                    "#
                ),
            )
            .main("main.tex")
            .position(1, 8)
            .test_completion(LatexArgumentCompletionProvider)
            .await;

        assert!(!actual_items.is_empty());
        assert_eq!(
            actual_items[0]
                .text_edit
                .as_ref()
                .and_then(|edit| edit.text_edit())
                .map(|edit| edit.range)
                .unwrap(),
            Range::new_simple(1, 8, 1, 11)
        );
    }

    #[tokio::test]
    async fn outside_mathbb_empty() {
        let actual_items = FeatureTester::new()
            .file(
                "main.tex",
                indoc!(
                    r#"
                        \usepackage{amsfonts}
                        \mathbb{}
                    "#
                ),
            )
            .main("main.tex")
            .position(1, 9)
            .test_completion(LatexArgumentCompletionProvider)
            .await;

        assert!(actual_items.is_empty());
    }
}
