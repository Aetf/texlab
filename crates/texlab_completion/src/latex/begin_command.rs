use super::combinators;
use crate::factory::{self, LatexComponentId};
use futures_boxed::boxed;
use texlab_protocol::{CompletionItem, CompletionParams};
use texlab_workspace::*;

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub struct LatexBeginCommandCompletionProvider;

impl FeatureProvider for LatexBeginCommandCompletionProvider {
    type Params = CompletionParams;
    type Output = Vec<CompletionItem>;

    #[boxed]
    async fn execute<'a>(&'a self, request: &'a FeatureRequest<Self::Params>) -> Self::Output {
        combinators::command(request, |_| {
            async move {
                let snippet = factory::command_snippet(
                    request,
                    "begin",
                    None,
                    "begin{$1}\n\t$0\n\\end{$1}",
                    &LatexComponentId::kernel(),
                );
                vec![snippet]
            }
        })
        .await
    }
}
