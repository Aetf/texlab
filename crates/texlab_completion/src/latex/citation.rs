use super::combinators::{self, ArgumentContext, Parameter};
use crate::factory;
use async_trait::async_trait;
use petgraph::graph::NodeIndex;
use texlab_feature::{Document, DocumentContent, FeatureProvider, FeatureRequest};
use texlab_protocol::{CompletionItem, CompletionParams, TextEdit};
use texlab_syntax::{bibtex, LANGUAGE_DATA};

#[derive(Debug, PartialEq, Eq, Clone, Copy, Default)]
pub struct LatexCitationCompletionProvider;

#[async_trait]
impl FeatureProvider for LatexCitationCompletionProvider {
    type Params = CompletionParams;
    type Output = Vec<CompletionItem>;

    async fn execute<'a>(&'a self, req: &'a FeatureRequest<Self::Params>) -> Self::Output {
        let parameters = LANGUAGE_DATA.citation_commands.iter().map(|cmd| Parameter {
            name: &cmd.name,
            index: cmd.index,
        });

        combinators::argument(req, parameters, |ctx| async move {
            let mut items = Vec::new();
            for doc in req.related() {
                if let DocumentContent::Bibtex(tree) = &doc.content {
                    for entry_node in tree.children(tree.root) {
                        if let Some(item) = Self::make_item(req, ctx, doc, tree, entry_node) {
                            items.push(item);
                        }
                    }
                }
            }
            items
        })
        .await
    }
}

impl LatexCitationCompletionProvider {
    fn make_item(
        req: &FeatureRequest<CompletionParams>,
        ctx: ArgumentContext,
        doc: &Document,
        tree: &bibtex::Tree,
        entry_node: NodeIndex,
    ) -> Option<CompletionItem> {
        let entry = tree.as_entry(entry_node)?;
        if entry.is_comment() {
            return None;
        }

        let key = entry.key.as_ref()?.text().to_owned();
        let text_edit = TextEdit::new(ctx.range, key.clone());
        let item = factory::citation(req, doc.uri.clone(), tree, entry_node, key, text_edit);
        Some(item)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;
    use texlab_feature::FeatureTester;
    use texlab_protocol::{CompletionTextEditExt, Range, RangeExt};

    #[tokio::test]
    async fn empty_latex_document() {
        let actual_items = FeatureTester::new()
            .file("main.tex", "")
            .main("main.tex")
            .position(0, 0)
            .test_completion(LatexCitationCompletionProvider)
            .await;

        assert!(actual_items.is_empty());
    }

    #[tokio::test]
    async fn empty_bibtex_document() {
        let actual_items = FeatureTester::new()
            .file("main.bib", "")
            .main("main.bib")
            .position(0, 0)
            .test_completion(LatexCitationCompletionProvider)
            .await;

        assert!(actual_items.is_empty());
    }

    #[tokio::test]
    async fn incomplete() {
        let actual_items = FeatureTester::new()
            .file(
                "main.tex",
                indoc!(
                    r#"
                        \addbibresource{main.bib}
                        \cite{
                        \begin{foo}
                        \end{bar}
                    "#
                ),
            )
            .file("main.bib", "@article{foo,}")
            .main("main.tex")
            .position(1, 6)
            .test_completion(LatexCitationCompletionProvider)
            .await;

        assert_eq!(actual_items.len(), 1);
        assert_eq!(actual_items[0].label, "foo");
        assert_eq!(
            actual_items[0]
                .text_edit
                .as_ref()
                .and_then(|edit| edit.text_edit())
                .map(|edit| edit.range)
                .unwrap(),
            Range::new_simple(1, 6, 1, 6)
        );
    }

    #[tokio::test]
    async fn empty_key() {
        let actual_items = FeatureTester::new()
            .file(
                "foo.tex",
                indoc!(
                    r#"
                        \addbibresource{bar.bib}
                        \cite{}  
                    "#
                ),
            )
            .file("bar.bib", "@article{foo,}")
            .file("baz.bib", "@article{bar,}")
            .main("foo.tex")
            .position(1, 6)
            .test_completion(LatexCitationCompletionProvider)
            .await;

        assert_eq!(actual_items.len(), 1);
        assert_eq!(actual_items[0].label, "foo");
        assert_eq!(
            actual_items[0]
                .text_edit
                .as_ref()
                .and_then(|edit| edit.text_edit())
                .map(|edit| edit.range)
                .unwrap(),
            Range::new_simple(1, 6, 1, 6)
        );
    }

    #[tokio::test]
    async fn single_key() {
        let actual_items = FeatureTester::new()
            .file(
                "foo.tex",
                indoc!(
                    r#"
                    \addbibresource{bar.bib}
                    \cite{foo}  
                "#
                ),
            )
            .file("bar.bib", "@article{foo,}")
            .file("baz.bib", "@article{bar,}")
            .main("foo.tex")
            .position(1, 6)
            .test_completion(LatexCitationCompletionProvider)
            .await;

        assert_eq!(actual_items.len(), 1);
        assert_eq!(actual_items[0].label, "foo");
        assert_eq!(
            actual_items[0]
                .text_edit
                .as_ref()
                .and_then(|edit| edit.text_edit())
                .map(|edit| edit.range)
                .unwrap(),
            Range::new_simple(1, 6, 1, 9)
        );
    }

    #[tokio::test]
    async fn second_key() {
        let actual_items = FeatureTester::new()
            .file(
                "foo.tex",
                indoc!(
                    r#"
                    \addbibresource{bar.bib}
                    \cite{foo,}  
                "#
                ),
            )
            .file("bar.bib", "@article{foo,}")
            .file("baz.bib", "@article{bar,}")
            .main("foo.tex")
            .position(1, 10)
            .test_completion(LatexCitationCompletionProvider)
            .await;

        assert_eq!(actual_items.len(), 1);
        assert_eq!(actual_items[0].label, "foo");
        assert_eq!(
            actual_items[0]
                .text_edit
                .as_ref()
                .and_then(|edit| edit.text_edit())
                .map(|edit| edit.range)
                .unwrap(),
            Range::new_simple(1, 10, 1, 10)
        );
    }

    #[tokio::test]
    async fn outside_cite() {
        let actual_items = FeatureTester::new()
            .file(
                "foo.tex",
                indoc!(
                    r#"
                        \addbibresource{bar.bib}
                        \cite{}  
                    "#
                ),
            )
            .file("bar.bib", "@article{foo,}")
            .file("baz.bib", "@article{bar,}")
            .main("foo.tex")
            .position(1, 7)
            .test_completion(LatexCitationCompletionProvider)
            .await;

        assert!(actual_items.is_empty());
    }
}
