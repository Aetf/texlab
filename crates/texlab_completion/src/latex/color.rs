use super::combinators::{self, Parameter};
use crate::factory;
use futures_boxed::boxed;
use texlab_feature::{FeatureProvider, FeatureRequest};
use texlab_protocol::{CompletionItem, CompletionParams, TextEdit};
use texlab_syntax::LANGUAGE_DATA;

#[derive(Debug, PartialEq, Eq, Clone, Copy, Default)]
pub struct LatexColorCompletionProvider;

impl FeatureProvider for LatexColorCompletionProvider {
    type Params = CompletionParams;
    type Output = Vec<CompletionItem>;

    #[boxed]
    async fn execute<'a>(&'a self, req: &'a FeatureRequest<Self::Params>) -> Self::Output {
        let parameters = LANGUAGE_DATA.color_commands.iter().map(|cmd| Parameter {
            name: &cmd.name,
            index: cmd.index,
        });

        combinators::argument(req, parameters, |ctx| async move {
            let mut items = Vec::new();
            for name in &LANGUAGE_DATA.colors {
                let text_edit = TextEdit::new(ctx.range, name.into());
                let item = factory::color(req, name, text_edit);
                items.push(item);
            }
            items
        })
        .await
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use texlab_feature::FeatureTester;
    use texlab_protocol::{Range, RangeExt};

    #[tokio::test]
    async fn empty_latex_document() {
        let actual_items = FeatureTester::new()
            .file("main.tex", "")
            .main("main.tex")
            .position(0, 0)
            .test_completion(LatexColorCompletionProvider)
            .await;

        assert!(actual_items.is_empty());
    }

    #[tokio::test]
    async fn empty_bibtex_document() {
        let actual_items = FeatureTester::new()
            .file("main.bib", "")
            .main("main.bib")
            .position(0, 0)
            .test_completion(LatexColorCompletionProvider)
            .await;

        assert!(actual_items.is_empty());
    }

    #[tokio::test]
    async fn inside_color() {
        let actual_items = FeatureTester::new()
            .file("main.tex", r#"\color{}"#)
            .main("main.tex")
            .position(0, 7)
            .test_completion(LatexColorCompletionProvider)
            .await;

        assert!(!actual_items.is_empty());
        assert_eq!(
            actual_items[0]
                .text_edit
                .as_ref()
                .map(|edit| edit.range)
                .unwrap(),
            Range::new_simple(0, 7, 0, 7)
        );
    }

    #[tokio::test]
    async fn inside_define_color_set() {
        let actual_items = FeatureTester::new()
            .file("main.tex", r#"\color{}"#)
            .main("main.tex")
            .position(0, 8)
            .test_completion(LatexColorCompletionProvider)
            .await;

        assert!(actual_items.is_empty());
    }
}
