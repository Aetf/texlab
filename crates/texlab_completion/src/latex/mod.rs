pub mod argument;
pub mod begin_command;
pub mod citation;
pub mod color;
pub mod color_model;
mod combinators;
pub mod component;
pub mod glossary;
pub mod import;
pub mod include;
pub mod label;
pub mod theorem;
pub mod tikz_lib;
pub mod user;
