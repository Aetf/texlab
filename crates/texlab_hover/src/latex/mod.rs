pub mod citation;
pub mod component;
pub mod include;
pub mod label;
pub mod preview;
