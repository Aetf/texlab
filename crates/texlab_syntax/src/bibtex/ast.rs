use crate::text::{Span, SyntaxNode};
use lsp_types::Range;

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum BibtexTokenKind {
    PreambleKind,
    StringKind,
    EntryKind,
    Word,
    Command,
    Assign,
    Comma,
    Concat,
    Quote,
    BeginBrace,
    EndBrace,
    BeginParen,
    EndParen,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct BibtexToken {
    pub span: Span,
    pub kind: BibtexTokenKind,
}

impl BibtexToken {
    pub fn new(span: Span, kind: BibtexTokenKind) -> Self {
        BibtexToken { span, kind }
    }

    pub fn text(&self) -> &str {
        &self.span.text
    }
}

impl SyntaxNode for BibtexToken {
    fn range(&self) -> Range {
        self.span.range
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct BibtexRoot {
    pub children: Vec<BibtexDeclaration>,
}

impl BibtexRoot {
    pub fn new(children: Vec<BibtexDeclaration>) -> Self {
        BibtexRoot { children }
    }
}

impl SyntaxNode for BibtexRoot {
    fn range(&self) -> Range {
        if self.children.is_empty() {
            Range::new_simple(0, 0, 0, 0)
        } else {
            Range::new(
                self.children[0].start(),
                self.children[self.children.len() - 1].end(),
            )
        }
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum BibtexDeclaration {
    Comment(Box<BibtexComment>),
    Preamble(Box<BibtexPreamble>),
    String(Box<BibtexString>),
    Entry(Box<BibtexEntry>),
}

impl BibtexDeclaration {
    pub fn accept<'a>(&'a self, visitor: &mut BibtexVisitor<'a>) {
        match self {
            BibtexDeclaration::Comment(comment) => visitor.visit_comment(comment),
            BibtexDeclaration::Preamble(preamble) => visitor.visit_preamble(preamble),
            BibtexDeclaration::String(string) => visitor.visit_string(string),
            BibtexDeclaration::Entry(entry) => visitor.visit_entry(entry),
        }
    }
}

impl SyntaxNode for BibtexDeclaration {
    fn range(&self) -> Range {
        match self {
            BibtexDeclaration::Comment(comment) => comment.range,
            BibtexDeclaration::Preamble(preamble) => preamble.range,
            BibtexDeclaration::String(string) => string.range,
            BibtexDeclaration::Entry(entry) => entry.range,
        }
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct BibtexComment {
    pub range: Range,
    pub token: BibtexToken,
}

impl BibtexComment {
    pub fn new(token: BibtexToken) -> Self {
        BibtexComment {
            range: token.range(),
            token,
        }
    }
}

impl SyntaxNode for BibtexComment {
    fn range(&self) -> Range {
        self.range
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct BibtexPreamble {
    pub range: Range,
    pub ty: BibtexToken,
    pub left: Option<BibtexToken>,
    pub content: Option<BibtexContent>,
    pub right: Option<BibtexToken>,
}

impl BibtexPreamble {
    pub fn new(
        ty: BibtexToken,
        left: Option<BibtexToken>,
        content: Option<BibtexContent>,
        right: Option<BibtexToken>,
    ) -> Self {
        let end = if let Some(ref right) = right {
            right.end()
        } else if let Some(ref content) = content {
            content.end()
        } else if let Some(ref left) = left {
            left.end()
        } else {
            ty.end()
        };
        BibtexPreamble {
            range: Range::new(ty.start(), end),
            ty,
            left,
            content,
            right,
        }
    }
}

impl SyntaxNode for BibtexPreamble {
    fn range(&self) -> Range {
        self.range
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct BibtexString {
    pub range: Range,
    pub ty: BibtexToken,
    pub left: Option<BibtexToken>,
    pub name: Option<BibtexToken>,
    pub assign: Option<BibtexToken>,
    pub value: Option<BibtexContent>,
    pub right: Option<BibtexToken>,
}

impl BibtexString {
    pub fn new(
        ty: BibtexToken,
        left: Option<BibtexToken>,
        name: Option<BibtexToken>,
        assign: Option<BibtexToken>,
        value: Option<BibtexContent>,
        right: Option<BibtexToken>,
    ) -> Self {
        let end = if let Some(ref right) = right {
            right.end()
        } else if let Some(ref value) = value {
            value.end()
        } else if let Some(ref assign) = assign {
            assign.end()
        } else if let Some(ref name) = name {
            name.end()
        } else if let Some(ref left) = left {
            left.end()
        } else {
            ty.end()
        };

        BibtexString {
            range: Range::new(ty.start(), end),
            ty,
            left,
            name,
            assign,
            value,
            right,
        }
    }
}

impl SyntaxNode for BibtexString {
    fn range(&self) -> Range {
        self.range
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct BibtexEntry {
    pub range: Range,
    pub ty: BibtexToken,
    pub left: Option<BibtexToken>,
    pub key: Option<BibtexToken>,
    pub comma: Option<BibtexToken>,
    pub fields: Vec<BibtexField>,
    pub right: Option<BibtexToken>,
}

impl BibtexEntry {
    pub fn new(
        ty: BibtexToken,
        left: Option<BibtexToken>,
        key: Option<BibtexToken>,
        comma: Option<BibtexToken>,
        fields: Vec<BibtexField>,
        right: Option<BibtexToken>,
    ) -> Self {
        let end = if let Some(ref right) = right {
            right.end()
        } else if !fields.is_empty() {
            fields[fields.len() - 1].range.end
        } else if let Some(ref comma) = comma {
            comma.end()
        } else if let Some(ref key) = key {
            key.end()
        } else if let Some(ref left) = left {
            left.end()
        } else {
            ty.end()
        };

        BibtexEntry {
            range: Range::new(ty.start(), end),
            ty,
            left,
            key,
            comma,
            fields,
            right,
        }
    }

    pub fn is_comment(&self) -> bool {
        self.ty.text().to_lowercase() == "@comment"
    }
}

impl SyntaxNode for BibtexEntry {
    fn range(&self) -> Range {
        self.range
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct BibtexField {
    pub range: Range,
    pub name: BibtexToken,
    pub assign: Option<BibtexToken>,
    pub content: Option<BibtexContent>,
    pub comma: Option<BibtexToken>,
}

impl BibtexField {
    pub fn new(
        name: BibtexToken,
        assign: Option<BibtexToken>,
        content: Option<BibtexContent>,
        comma: Option<BibtexToken>,
    ) -> Self {
        let end = if let Some(ref comma) = comma {
            comma.end()
        } else if let Some(ref content) = content {
            content.end()
        } else if let Some(ref assign) = assign {
            assign.end()
        } else {
            name.end()
        };

        BibtexField {
            range: Range::new(name.start(), end),
            name,
            assign,
            content,
            comma,
        }
    }
}

impl SyntaxNode for BibtexField {
    fn range(&self) -> Range {
        self.range
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum BibtexContent {
    Word(BibtexWord),
    Command(BibtexCommand),
    QuotedContent(BibtexQuotedContent),
    BracedContent(BibtexBracedContent),
    Concat(Box<BibtexConcat>),
}

impl BibtexContent {
    pub fn accept<'a>(&'a self, visitor: &mut BibtexVisitor<'a>) {
        match self {
            BibtexContent::Word(word) => visitor.visit_word(word),
            BibtexContent::Command(command) => visitor.visit_command(command),
            BibtexContent::QuotedContent(content) => visitor.visit_quoted_content(content),
            BibtexContent::BracedContent(content) => visitor.visit_braced_content(content),
            BibtexContent::Concat(concat) => visitor.visit_concat(concat),
        }
    }
}

impl SyntaxNode for BibtexContent {
    fn range(&self) -> Range {
        match self {
            BibtexContent::Word(word) => word.range(),
            BibtexContent::Command(command) => command.range(),
            BibtexContent::QuotedContent(content) => content.range(),
            BibtexContent::BracedContent(content) => content.range(),
            BibtexContent::Concat(concat) => concat.range(),
        }
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct BibtexWord {
    pub range: Range,
    pub token: BibtexToken,
}

impl BibtexWord {
    pub fn new(token: BibtexToken) -> Self {
        BibtexWord {
            range: token.range(),
            token,
        }
    }
}

impl SyntaxNode for BibtexWord {
    fn range(&self) -> Range {
        self.range
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct BibtexCommand {
    pub range: Range,
    pub token: BibtexToken,
}

impl BibtexCommand {
    pub fn new(token: BibtexToken) -> Self {
        BibtexCommand {
            range: token.range(),
            token,
        }
    }
}

impl SyntaxNode for BibtexCommand {
    fn range(&self) -> Range {
        self.range
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct BibtexQuotedContent {
    pub range: Range,
    pub left: BibtexToken,
    pub children: Vec<BibtexContent>,
    pub right: Option<BibtexToken>,
}

impl BibtexQuotedContent {
    pub fn new(
        left: BibtexToken,
        children: Vec<BibtexContent>,
        right: Option<BibtexToken>,
    ) -> Self {
        let end = if let Some(ref right) = right {
            right.end()
        } else if !children.is_empty() {
            children[children.len() - 1].end()
        } else {
            left.end()
        };

        BibtexQuotedContent {
            range: Range::new(left.start(), end),
            left,
            children,
            right,
        }
    }
}

impl SyntaxNode for BibtexQuotedContent {
    fn range(&self) -> Range {
        self.range
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct BibtexBracedContent {
    pub range: Range,
    pub left: BibtexToken,
    pub children: Vec<BibtexContent>,
    pub right: Option<BibtexToken>,
}

impl BibtexBracedContent {
    pub fn new(
        left: BibtexToken,
        children: Vec<BibtexContent>,
        right: Option<BibtexToken>,
    ) -> Self {
        let end = if let Some(ref right) = right {
            right.end()
        } else if !children.is_empty() {
            children[children.len() - 1].end()
        } else {
            left.end()
        };

        BibtexBracedContent {
            range: Range::new(left.start(), end),
            left,
            children,
            right,
        }
    }
}

impl SyntaxNode for BibtexBracedContent {
    fn range(&self) -> Range {
        self.range
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct BibtexConcat {
    pub range: Range,
    pub left: BibtexContent,
    pub operator: BibtexToken,
    pub right: Option<BibtexContent>,
}

impl BibtexConcat {
    pub fn new(left: BibtexContent, operator: BibtexToken, right: Option<BibtexContent>) -> Self {
        let end = if let Some(ref right) = right {
            right.end()
        } else {
            operator.end()
        };

        BibtexConcat {
            range: Range::new(left.start(), end),
            left,
            operator,
            right,
        }
    }
}

impl SyntaxNode for BibtexConcat {
    fn range(&self) -> Range {
        self.range
    }
}

pub trait BibtexVisitor<'a> {
    fn visit_root(&mut self, root: &'a BibtexRoot);

    fn visit_comment(&mut self, comment: &'a BibtexComment);

    fn visit_preamble(&mut self, preamble: &'a BibtexPreamble);

    fn visit_string(&mut self, string: &'a BibtexString);

    fn visit_entry(&mut self, entry: &'a BibtexEntry);

    fn visit_field(&mut self, field: &'a BibtexField);

    fn visit_word(&mut self, word: &'a BibtexWord);

    fn visit_command(&mut self, command: &'a BibtexCommand);

    fn visit_quoted_content(&mut self, content: &'a BibtexQuotedContent);

    fn visit_braced_content(&mut self, content: &'a BibtexBracedContent);

    fn visit_concat(&mut self, concat: &'a BibtexConcat);
}

pub struct BibtexWalker;

impl BibtexWalker {
    pub fn walk_root<'a>(visitor: &mut BibtexVisitor<'a>, root: &'a BibtexRoot) {
        for declaration in &root.children {
            declaration.accept(visitor);
        }
    }

    pub fn walk_preamble<'a>(visitor: &mut BibtexVisitor<'a>, preamble: &'a BibtexPreamble) {
        if let Some(ref content) = preamble.content {
            content.accept(visitor);
        }
    }

    pub fn walk_string<'a>(visitor: &mut BibtexVisitor<'a>, string: &'a BibtexString) {
        if let Some(ref value) = string.value {
            value.accept(visitor);
        }
    }

    pub fn walk_entry<'a>(visitor: &mut BibtexVisitor<'a>, entry: &'a BibtexEntry) {
        for field in &entry.fields {
            visitor.visit_field(field);
        }
    }

    pub fn walk_field<'a>(visitor: &mut BibtexVisitor<'a>, field: &'a BibtexField) {
        if let Some(ref content) = field.content {
            content.accept(visitor);
        }
    }

    pub fn walk_quoted_content<'a>(
        visitor: &mut BibtexVisitor<'a>,
        content: &'a BibtexQuotedContent,
    ) {
        for child in &content.children {
            child.accept(visitor);
        }
    }

    pub fn walk_braced_content<'a>(
        visitor: &mut BibtexVisitor<'a>,
        content: &'a BibtexBracedContent,
    ) {
        for child in &content.children {
            child.accept(visitor);
        }
    }

    pub fn walk_concat<'a>(visitor: &mut BibtexVisitor<'a>, concat: &'a BibtexConcat) {
        concat.left.accept(visitor);
        if let Some(ref right) = concat.right {
            right.accept(visitor);
        }
    }
}
