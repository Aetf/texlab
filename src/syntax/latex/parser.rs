use crate::syntax::latex::ast::*;
use std::iter::Peekable;

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
enum LatexScope {
    Root,
    Group,
    Options,
    Math,
}

pub struct LatexParser<I: Iterator<Item = LatexToken>> {
    tokens: Peekable<I>,
}

impl<I: Iterator<Item = LatexToken>> LatexParser<I> {
    pub fn new(tokens: I) -> Self {
        LatexParser {
            tokens: tokens.peekable(),
        }
    }

    pub fn root(&mut self) -> LatexRoot {
        let children = self.content(LatexScope::Root);
        LatexRoot::new(children)
    }

    fn content(&mut self, scope: LatexScope) -> Vec<LatexContent> {
        let mut children = Vec::new();
        while let Some(ref token) = self.tokens.peek() {
            match token.kind {
                LatexTokenKind::Word | LatexTokenKind::BeginOptions => {
                    let text = Box::new(self.text(scope));
                    children.push(LatexContent::Text(text));
                }
                LatexTokenKind::Command => {
                    let command = Box::new(self.command());
                    children.push(LatexContent::Command(command));
                }
                LatexTokenKind::Math => {
                    if scope == LatexScope::Math {
                        return children;
                    } else {
                        let group = Box::new(self.group(LatexGroupKind::Math));
                        children.push(LatexContent::Group(group));
                    }
                }
                LatexTokenKind::BeginGroup => {
                    let group = Box::new(self.group(LatexGroupKind::Group));
                    children.push(LatexContent::Group(group));
                }
                LatexTokenKind::EndGroup => {
                    if scope == LatexScope::Root {
                        self.tokens.next();
                    } else {
                        return children;
                    }
                }
                LatexTokenKind::EndOptions => {
                    if scope == LatexScope::Options {
                        return children;
                    } else {
                        let text = Box::new(self.text(scope));
                        children.push(LatexContent::Text(text));
                    }
                }
            }
        }
        children
    }

    fn command(&mut self) -> LatexCommand {
        let name = self.tokens.next().unwrap();
        let options = if self.next_of_kind(LatexTokenKind::BeginOptions) {
            Some(self.group(LatexGroupKind::Options))
        } else {
            None
        };

        let mut args = Vec::new();
        while self.next_of_kind(LatexTokenKind::BeginGroup) {
            args.push(self.group(LatexGroupKind::Group));
        }

        LatexCommand::new(name, options, args)
    }

    fn group(&mut self, kind: LatexGroupKind) -> LatexGroup {
        let left = self.tokens.next().unwrap();
        let scope = match kind {
            LatexGroupKind::Group => LatexScope::Group,
            LatexGroupKind::Options => LatexScope::Options,
            LatexGroupKind::Math => LatexScope::Math,
        };
        let children = self.content(scope);
        let right_kind = match kind {
            LatexGroupKind::Group => LatexTokenKind::EndGroup,
            LatexGroupKind::Options => LatexTokenKind::EndOptions,
            LatexGroupKind::Math => LatexTokenKind::Math,
        };

        let right = if self.next_of_kind(right_kind) {
            self.tokens.next()
        } else {
            None
        };

        LatexGroup::new(left, children, right, kind)
    }

    fn text(&mut self, scope: LatexScope) -> LatexText {
        let mut words = Vec::new();
        while let Some(ref token) = self.tokens.peek() {
            let kind = token.kind;
            let opts = kind == LatexTokenKind::EndOptions && scope != LatexScope::Options;
            if kind == LatexTokenKind::Word || kind == LatexTokenKind::BeginOptions || opts {
                words.push(self.tokens.next().unwrap());
            } else {
                break;
            }
        }
        LatexText::new(words)
    }

    fn next_of_kind(&mut self, kind: LatexTokenKind) -> bool {
        if let Some(ref token) = self.tokens.peek() {
            token.kind == kind
        } else {
            false
        }
    }
}
