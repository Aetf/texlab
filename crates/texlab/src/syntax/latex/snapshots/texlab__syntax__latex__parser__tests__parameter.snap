---
source: src/syntax/latex/parser.rs
expression: "setup(r#\"#1\"#)"

---
ROOT@0..2
  PREAMBLE@0..2
    TEXT@0..2
      WORD@0..2 "#1"

